src/gui/qgsbrowserdockwidget.cpp

```diff
@@ -197,7 +197,18 @@ void QgsBrowserDockWidget::showContextMenu( QPoint pt )
     menu->addAction( tr( "Add a Directory..." ), this, SLOT( addFavoriteDirectory() ) );
   }
 
+  const QList<QMenu *> menus = item->menus();
   QList<QAction *> actions = item->actions();
+
+  if ( !menus.isEmpty() )
+  {
+    for ( QMenu *mn : menus )
+    {
+      mn->setParent( this );
```

Comment:
The parent should be `menu`. Parenting to this means the menu will exist for the life of the dock widget (i.e. forever!)

Reply:
it appears that the menu is showing not like what we are expecting if we set the parent manually. I decided to just use `addMenu` method with `QString` signature which will take care of the ownership. http://doc.qt.io/qt-5/qmenu.html#addMenu what do you think?

Reply:
What happens when you reparent it? The current method is leaky because the new menus created by calling menus() are never deleted.

Reply:
the submenu that I created got stacked with the main context menu :sweat:  but seems like it works after we pass the parent into the `menus()` parameter just like what @m-kuhn suggested.